File: LeetCode/Python/0977_Squares_of_a_Sorted_Array#2.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        
        right = 0
        for num in nums:
            if num >= 0:
                break
            right += 1
        
        left = right - 1
        squares = []
        
        while left >= 0 and right < n:
            if nums[left] ** 2 > nums[right] ** 2:
                squares.append(nums[right] ** 2)
                right += 1
            else:
                squares.append(nums[left] ** 2)
                left -= 1
                
        while left >= 0:
            squares.append(nums[left] ** 2)
            left -= 1
            
        while right < n:
            squares.append(nums[right] ** 2)
            right += 1
                
        return squares

File: LeetCode/Java/0977_Squares_of_a_Sorted_Array_#2.java
class Solution {
    public int[] sortedSquares(int[] nums) {
        int n = nums.length;
        int right = 0;
        for (int num: nums) {
            if (num >= 0)
                break;
            right++;
        }
        
        int left = right - 1;
        int[] squares = new int[n];
        int index = 0;
        
        while (left >= 0 && right < n) {
            if (nums[left] * nums[left] < nums[right] * nums[right])
                squares[index++] = nums[left] * nums[left--];
            else
                squares[index++] = nums[right] * nums[right++];
        }
        
        while (left >= 0)
            squares[index++] = nums[left] * nums[left--];
        
        while (right < n)
            squares[index++] = nums[right] * nums[right++];
        
        return squares;
    }
}
